### DISCUSSION TITLE
Red Black Tree

### DISCUSSION THREAD

**Discussion by vumrra** [subak-ba/cs-study#49] [자료 구조]
# Red Black Tree

- BST의 한 종류이다.
- 스스로 균형을 유지하는 트리이다.
- BST의 worst case의 단점을 개선한다. (한쪽으로 치우친 트리)
- 모든 노드는 red or black을 가진다.

### 2. 루트 노드는 black이다.

nil노드란?

- null을 의미하는 노드이다.
- 자식 노드가 없을 때 자식 노드 위치에 nil노드로 표기한다.
- nil노드는 일반적인 노드와 같은 취급을 한다.
- red black tree에서는 모든 leaf노드는 nil노드이다.

### 5. 임의의 노드에서 nil 노드들 까지 가는 경로들의 black 수는 같다. (자기 자신 제외)

![image](https://github.com/subak-ba/cs-study/assets/131235625/531eb508-9a87-40b1-9427-0ad9bb33c53e)

black height

- 임의의 노드에서의 nil노드 까지의 경로에 있는 black의 수

red black tree가 조건 5를 만족하고 있고 두 자식 노드가 같은 색일때 부모와 두 자식의 색을 바꿔도 조건 5를 만족한다.

## 균형을 유지하는 법

- 삽입과 삭제시 주로 4, 5번 조건이 위반된다.
- 이를 해결하기 위해 트리의 구조를 변경하여 트리의 균형을 유지한다.

삽입 → 조건 위반 여부 확인 → 위반 되었다면? 재조정

## 삽입

- 삽입하는 노드의 색은 항상 red
    - 조건 5는 임의의 노드에서 nil 노드들 까지 가는 경로들의 black 수는 같다라는 조건을 만족하기 위해

→ insert(10)

![image](https://github.com/subak-ba/cs-study/assets/131235625/c8295e4e-d70a-4e93-8d48-14ebc37c2bb9)

- 현재 루트 노드가 red 이므로 조건 2 위반 → 루트 노드 black으로 변경

> nil 노드는 생략 하겠습니다.
> 

## case. 0

> 삽입되는 노드의 부모가 black일 경우
> 
- 새로운 노드(red)가 삽입된다고 해서 RBT 조건에 어긋나는 것이 없기 때문에 OK이다.

## case. 1, 2

> 삽입되는 노드의 부모가 red이고 삼촌 노드가 black인 경우
> 

> 삽입된 자식이 부모 노드의 왼쪽 자식 → case 1, 오른쪽 자식 → case 2
> 

![image](https://github.com/subak-ba/cs-study/assets/131235625/1f19168d-131c-4d34-92a2-ee382391b3cd)

- 이렇게 red 노드가 연속적으로 존재하는(double red) 상황이 발생한다. → 4번 규칙 위반
- 해결하기 위해서는 BST의 조건도 만족하면 서 RBT의 조건도 만족하여야한다.

→ 이를 해결하기 위해 부모 노드를 기준으로 오른쪽으로 로테이트 시킨다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/f9db1198-1d31-48ee-a9de-225455ea044b)

→ 회전 후 부모와 조부모 노드의 색을 변경한다. → double red 상황 해결

## case. 3

> 부모 노드와 삼촌 노드가 모두 red인 경우

![image](https://github.com/subak-ba/cs-study/assets/131235625/c7a49ab7-b147-4aa6-9784-5372d2815797)

- double red 발생 → 조건 4 위반
- 삽입한 노드 기준으로 조부모 노드부터 nil 노드들의 경로의 black 노드의 수가 같지 않다 → 조건 5 위반

부모 노드와 삼촌 노드를 black으로 바꾼 후 조무보를 red로 바꾸면 트리 전체의 black height를 유지할 수 있다.

<img width="619" alt="스크린샷 2024-04-18 오전 11 26 32" src="https://github.com/subak-ba/cs-study/assets/131235625/1a7abd1f-d577-4b04-abf7-30f7455d3357">

이후 조부모 노드 위에 red 노드가 존재하여 double red 상황이 또 일어날 수 있기 때문에 조부모 노드에 대해 다시 case 0 부터 검사를 진행한다.

## 시간 복잡도

탐색 - avg, worst O(log n)

삽입, 삭제 - avg, worst O(log n)

## AVL.T vs RB.T

- AVL tree
    - 더 엄격한 규정, RB.T보다 더 빠른 조회 성능
    - 각 노드별로 BF를 저장하므로 노드마다 4Byte 만큼의 저장 공간이 필요한 단점이 있다
    - 더 빠른 검색이 필요한 경우 사용
- RB.T
    - 상대적으로 느슨한 균형이다 (두 서브트리의 높이 차이가 최대 2배가 난다)
    - 로테이트가 거의 이루어지지 않는다
    - black or red만 저장하면 되므로 노드당 1비트면 된다
    - 삽입, 삭제 성능이 AVL.T 보다 빠르다 (AVL는 삽.삭시BF를 확인하기 위해 루트 노드까지 올라간다.)

**Comment by onee-only** [comment]
delete에 대한 케이스도 추가 해 주실 수 있나요? 자료 찾아보니 많은 경우의 수가 있는 것 같아서 간단한 케이스 몇개라도 있으면 좋을 것 같아요

**Reply by vumrra** [reply]
넵 삭제 case도 추가하도록 하겠습니다

**Comment by onee-only** [comment]
insert case 3에서

> 부모 노드와 삼촌 노드를 black으로 바꾼 후 조무보를 red로 바꾸면 트리 전체의 black height를 유지할 수 있다.

바꾼 후의 사진 자료에 부모 노드만 black으로 바뀌어 있는 것 같아요!

**Reply by vumrra** [reply]
확인해주셔서 감사합니다 예제 사진을 변경하였습니다!

**Reply by onee-only** [reply]
case 3에서 둘 다 black이게 되면 조건 5번을 위반하는 것 같아요
